pkg/verify/verify.go

```diff
@@ -0,0 +1,203 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package verify
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/tlog"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+	"github.com/transparency-dev/merkle/proof"
+	"github.com/transparency-dev/merkle/rfc6962"
+)
+
+func ProveConsistency(ctx context.Context, rClient *client.Rekor,
+	oldStateHash []byte, oldTreeSize int64, sth util.SignedCheckpoint, treeID string) error {
+	switch {
+	case oldTreeSize == int64(sth.Size):
+		if !bytes.Equal(oldStateHash, sth.Hash) {
+			return errors.New("old root hash does not match STH hash")
+		}
+	case oldTreeSize < int64(sth.Size):
+		consistencyParams := tlog.NewGetLogProofParamsWithContext(ctx)
+		consistencyParams.FirstSize = &oldTreeSize   // Root hash at the time the proof was returned
+		consistencyParams.LastSize = int64(sth.Size) // Root hash verified with rekor pubkey
+		consistencyParams.TreeID = &treeID
+		consistencyProof, err := rClient.Tlog.GetLogProof(consistencyParams)
+		if err != nil {
+			return err
+		}
+		var hashes [][]byte
+		for _, h := range consistencyProof.Payload.Hashes {
+			b, err := hex.DecodeString(h)
+			if err != nil {
+				return errors.New("error decoding consistency proof hashes")
+			}
+			hashes = append(hashes, b)
+		}
+		if err := proof.VerifyConsistency(rfc6962.DefaultHasher,
+			uint64(oldTreeSize), sth.Size, hashes, oldStateHash, sth.Hash); err != nil {
+			return err
+		}
+	case oldTreeSize > int64(sth.Size):
+		return errors.New("inclusion proof returned a tree size larger than the verified tree size")
+	}
+	return nil
+
+}
+
+// VerifyRootHash verifies the provided root hash at the treeSize by verifying consistenct
+// against an STH.
+//nolint
+func VerifyRootHash(ctx context.Context, rClient *client.Rekor, verifier signature.Verifier,
```

Comment:
See comment below, but I think this function still has value, just used differently. I could see this used by witnesses to calculate consistency from a trusted state (`rootHash`) to current. a nit, maybe a different function name, like VerifyCurrentCheckpoint?

Reply:
I think done as above.